Detectors/TPC/reconstruction/src/TPCFastTransformHelperO2.cxx

```diff
@@ -0,0 +1,312 @@
+    distortion.setTPCrow(iRow, xRow, nPads, padWidth, 0);
+  }
+
+  // set some initial values, will be reinitialised later int updateCalibration()
+  const double t0 = 0.;
+  const double vdCorrY = 0.;
+  const double ldCorr = 0.;
+  const double tpcAlignmentZ = 0.;
+  const double tofCorr = 0.;
+  const double primVtxZ = 0.;
+  const double initTimeStamp = -1;
+  fastTransform.setCalibration(initTimeStamp, t0, vDrift, vdCorrY, ldCorr, tofCorr, primVtxZ, tpcAlignmentZ);
+
+  IrregularSpline2D3D spline;
+  {
+    int nKnotsU = 15;
+    int nAxisTicksU = mapper.getNumberOfPadsInRowSector(10);
+    int nKnotsV = 20;
+    int nAxisTicksV = lastTimeBin + 1;
+    float knotsU[nKnotsU];
+    float knotsV[nKnotsV];
+    for (int i = 0; i < nKnotsU; i++)
+      knotsU[i] = 1. / (nKnotsU - 1) * i;
+    for (int i = 0; i < nKnotsV; i++)
+      knotsV[i] = 1. / (nKnotsV - 1) * i;
+
+    // TODO: adjust the grid
+
+    double d1 = 0.6;
+    double d2 = 0.9 - d1;
+    double d3 = 1. - d2 - d1;
+
+    for (int i = 0; i < 5; i++) { // 5 bins in first 6% of drift
+      knotsV[i] = i / 4. * d1;
+    }
+    for (int i = 0; i < 10; i++) { // 10 bins for 6% <-> 90%
+      knotsV[4 + i] = d1 + i / 9. * d2;
+    }
+    for (int i = 0; i < 5; i++) { // 5 bins for last 90% <-> 100%
+      knotsV[13 + i] = d1 + d2 + i / 4. * d3;
+    }
+
+    spline.construct(nKnotsU, knotsU, nAxisTicksU,
+                     nKnotsV, knotsV, nAxisTicksV);
+  }
+  distortion.setApproximationScenario(0, spline);
+  distortion.finishConstruction();
+  fastTransform.finishConstruction();
+
+  // check if calculated pad geometry is consistent with the map
+  testGeometry(fastTransform);
+
+  updateCalibration(fastTransform, TimeStamp);
+
+  return std::move(fastTransformPtr);
+}
+
+int TPCFastTransformHelperO2::updateCalibration(ali_tpc_common::tpc_fast_transformation::TPCFastTransform& fastTransform, Long_t TimeStamp)
+{
+  // Update the calibration with the new time stamp
+
+  if (!mIsInitialized) {
+    init();
+  }
+
+  Long_t lastTS = fastTransform.getTimeStamp();
+
+  // deinitialize
+
+  fastTransform.setTimeStamp(-1);
+
+  if (TimeStamp < 0) {
+    return 0;
+  }
+
+  // search for the calibration database ...
+
+  const static ParameterDetector& detParam = ParameterDetector::defaultInstance();
+  const static ParameterGas& gasParam = ParameterGas::defaultInstance();
+  const static ParameterElectronics& elParam = ParameterElectronics::defaultInstance();
+
+  // calibration found, set the initialized status back
+
+  fastTransform.setTimeStamp(lastTS);
+
+  // less than 60 seconds from the previois time stamp, don't do anything
+
+  if (lastTS >= 0 && TMath::Abs(lastTS - TimeStamp) < 60) {
+    return 0;
+  }
+
+  // start the initialization
+
+  fastTransform.setTimeStamp(TimeStamp);
+
+  // find last calibrated time bin
+
+  const double vDrift = elParam.getZBinWidth() * gasParam.getVdrift();
+
+  //mLastTimeBin = detParam.getTPClength() / vDrift  + 1;
```

Comment:
I think it's not the task of the `update` method to recalculate the last time bin, so this can be removed completely